File: Python/treehopper/libraries/SpiDevice.py
from treehopper.api import Spi
from treehopper.api.Spi import ChipSelectMode, SpiMode, SpiBurstMode
from treehopper.api.Pin import SpiChipSelectPin


class SpiDevice:
    """
    SPI device

    This class provides a simple wrapper around Spi.send_receive() that preserves the chip-select, mode, and frequency
    configurations.
    """
    def __init__(self, spi_module: Spi, chip_select: SpiChipSelectPin, chip_select_mode=ChipSelectMode.SpiActiveLow, speed_mhz=6, spi_mode=SpiMode.Mode00):
        """
        Construct a new SPI device.
        :param spi_module: The SPI module to use
        :param chip_select: The chip select pin to use
        :param chip_select_mode: The chip select mode to use (e.g., ChipSelectMode.SpiActiveLow)
        :param speed_mhz: The speed, in MHz, to use.
        :param spi_mode: The SPI mode (e.g., SpiMode.Mode00) to use.

        Due to an open firmware issue, we recommend avoiding frequencies between 0.8 and 6 MHz, as these can cause ISR glitches under certain bus constraints.
        """
        self.chip_select_mode = chip_select_mode
        self.chip_select = chip_select
        self._spi = spi_module  # type: Spi
        self.frequency = speed_mhz
        self.mode = spi_mode
        self._spi.enabled = True
        chip_select.make_digital_push_pull_out()

        if chip_select_mode == ChipSelectMode.PulseLowAtBeginning or chip_select_mode == ChipSelectMode.PulseLowAtEnd or chip_select_mode == ChipSelectMode.SpiActiveLow:
            chip_select.digital_value = True
        else:
            chip_select.digital_value = False

    def send_receive(self, data_to_send, burst=SpiBurstMode.NoBurst):
        """
        Send and/or receive data using the settings specified by the constructor.
        :param data_to_send: A list-like object of data to send
        :param burst: the SPI burst mode (choose from SpiBurstMode members).
        :return: Data received during the transaction.
        """
        self._spi.send_receive(data_to_send, self.chip_select, self.chip_select_mode, self.frequency, burst, self.mode)

File: NET/API/Treehopper.Libraries/SpiDevice.cs
using System.Threading.Tasks;

namespace Treehopper.Libraries
{
    /// <summary>
    ///     Represents a peripheral attached to the SPI bus
    /// </summary>
    public class SpiDevice
    {
        private readonly Spi _spi;

        /// <summary>
        ///     Construct an SPI device attached to a particular module
        /// </summary>
        /// <param name="spiModule">The module this device is attached to</param>
        /// <param name="chipSelect">The chip select pin used by this device</param>
        /// <param name="chipSelectMode">The ChipSelectMode to use with this device</param>
        /// <param name="speedMhz">The speed to operate this device at</param>
        /// <param name="mode">The SpiMode of this device</param>
        public SpiDevice(Spi spiModule, SpiChipSelectPin chipSelect,
            ChipSelectMode chipSelectMode = ChipSelectMode.SpiActiveLow, double speedMhz = 6,
            SpiMode mode = SpiMode.Mode00)
        {
            ChipSelectMode = chipSelectMode;
            ChipSelect = chipSelect;
            _spi = spiModule;
            Frequency = speedMhz;
            Mode = mode;

            _spi.Enabled = true;

            ChipSelect.MakeDigitalPushPullOutAsync();

            // Set the initial chip select state
            if (chipSelectMode == ChipSelectMode.PulseLowAtBeginning ||
                chipSelectMode == ChipSelectMode.PulseLowAtEnd || chipSelectMode == ChipSelectMode.SpiActiveLow)
                ChipSelect.DigitalValue = true;
            else
                ChipSelect.DigitalValue = false;
        }

        /// <summary>
        ///     Get or set the pin to use for chip-select duties.
        /// </summary>
        /// <remarks>
        ///     Almost every SPI peripheral chip has some sort of chip select (which may be called load, strobe, or enable,
        ///     depending on the type of chip). You can use any <see cref="Pin" /> for chip-select duties as long as it belongs to
        ///     the same board as this SPI peripheral (i.e., you can't use a pin from one Treehopper as a chip-select for the SPI
        ///     port on another Treehopper).
        ///     Chip-selects are controlled at the firmware, not peripheral, level, which offers quite a bit of flexibility in
        ///     choosing the behavior. Make sure to set <see cref="ChipSelectMode" /> properly for your device.
        /// </remarks>
        public SpiChipSelectPin ChipSelect { get; protected set; }

        /// <summary>
        ///     Get or set the SPI module's mode
        /// </summary>
        /// <remarks>
        ///     The SPI module supports the four SPI modes: 00, 01, 10, 11. See <see cref="SpiMode" /> for more info.
        /// </remarks>
        public SpiMode Mode { get; protected set; }

        /// <summary>
        ///     Gets or sets the Frequency, in MHz, that this device will use for SPI communication.
        /// </summary>
        /// <remarks>
        ///     The SPI module can operate from 0.09375 MHz (93.75 kHz) to 24 MHz. Setting Frequency outside of those limits will
        ///     result in clipping, plus a debug notice.
        ///     Note that SPI transfers are managed by the CPU, as Treehopper's MCU has no DMA. As a result, there are diminishing
        ///     performance returns above 8 MHz, as dead space starts appearing between bytes as the SPI module waits for the CPU
        ///     to push the next byte into the register.
        /// </remarks>
        public double Frequency { get; set; }

        /// <summary>
        ///     The chip select mode to use with transactions
        /// </summary>
        public ChipSelectMode ChipSelectMode { get; protected set; }

        /// <summary>
        ///     Start an SPI transaction
        /// </summary>
        /// <param name="dataToSend">The data to send</param>
        /// <param name="burst">The burst mode, if any, to use</param>
        /// <returns>Data received by the peripheral</returns>
        public Task<byte[]> SendReceiveAsync(byte[] dataToSend, SpiBurstMode burst = SpiBurstMode.NoBurst)
        {
            return _spi.SendReceiveAsync(dataToSend, ChipSelect, ChipSelectMode, Frequency, burst, Mode);
        }
    }
}